mlc_llm/quantization/smoothquant_utils.py

```diff
@@ -0,0 +1,322 @@
+        scale_params[f"sq_scale_{idx+1}"] = tvm.nd.array(np.array([w_scale]), dev)
+        idx += 2
+
+    return scale_params
+
+
+def _smooth(
+    mod: tvm.IRModule,
+    params: List[tvm.nd.NDArray],
+    funcs: List[str],
+    dataset: List[tvm.nd.NDArray],
+    config: Dict[str, Any],
+):
+    mod = mlc_llm.transform.SmoothQuantAnnotator()(mod)
+    stat_mod = mlc_llm.transform.SmoothQuantStatCollector()(mod)
+    stat_mod = mlc_llm.transform.FuseTransposeMatmul()(stat_mod)
+
+    prefill, decode, kvc, _, _ = get_runtime_func(funcs, stat_mod)
+
+    # Calculate max statistics
+    # Number of dimension in a_stat/w_stat is equal to 3, where:
+    #  * 1st dimension - number of outputs in compute graph / 2
+    #  * 2nd dimension - number of elements in dataset
+    #  * 3rd dimension - scale(multiplier) dimension.
+    a_stat = None
+    w_stat = None
+
+    target = tvm.target.Target.current(allow_none=False)
+    for data in tqdm(dataset, desc="Smoothing"):
+        # Create KV-cache
+        kv_caches = kvc()
+        num_tokens = data.shape[1]
+        seq_len_shape = tvm.runtime.ShapeTuple([num_tokens])
+
+        # Run Encoder and update statistics for activations/weights
+        (logits, kv_caches), outputs = prefill(data, seq_len_shape, kv_caches, *params)
+        a_stat = _accumulate_act_outlier_stat(a_stat, outputs)
+        w_stat = _accumulate_weight_outlier_stat(w_stat, outputs)
+
+        # Run Decoder and update statistics for activations/weights
+        for _ in range(config["decoder_invoke_num"]):
+            # TODO: support softmax with temperature.
+            logits_max = np.argmax(logits.numpy(), axis=-1).astype("int32")
+            if logits_max[0] in config["stop_tokens"]:
+                break
+            next_token = tvm.nd.array(logits_max, device=tvm.device(target.kind.default_keys[0]))
+            num_tokens += logits_max.shape[1]
+            seq_len_shape = tvm.runtime.ShapeTuple([num_tokens])
+            (logits, kv_caches), outputs = decode(next_token, seq_len_shape, kv_caches, *params)
+            a_stat = _accumulate_act_outlier_stat(a_stat, outputs)
+            w_stat = _accumulate_weight_outlier_stat(w_stat, outputs)
+
+    a_stat = [np.max(s, axis=0) for s in a_stat]
+    w_stat = [np.max(s, axis=0) for s in w_stat]
+    # Use the same statistics for "prefill"/"decode"
+    stat = dict.fromkeys(funcs)
+    stat["prefill"] = (a_stat, w_stat)
+    stat["decode"] = (a_stat, w_stat)
+    for fname in funcs:
+        scale_params = _calculate_scale_params(fname, stat, config, tvm.cpu(0))
+        mod = relax.transform.BindParams(fname, scale_params)(mod)
+
+    mod = mlc_llm.transform.SmoothQuantOpConverter("multiply")(mod)
+    return mod
+
+
+def _calibrate(
+    mod: tvm.IRModule,
+    params: List[tvm.nd.NDArray],
+    funcs: List[str],
+    dataset: List[tvm.nd.NDArray],
+    config: Dict[str, Any],
+):
+    mod = mlc_llm.transform.SmoothQuantAnnotator("quantize")(mod)
```

Comment:
The input `mod` already has `smooth` op right? Why do we need to annotate again?

Reply:
Yes, You are right, we do have `smooth` ops here. But these `smooth `ops operate like `multiply `op for activations smoothing. Here we annotate IRModule with new `smooth `ops that will be converted into analogue of `qnn.quantize` op (it should be qnn.quantize, but now this op is absent, that's why I use smooth op with attribute mode="quantize")

Reply:
I see. Rather than having a chain of `smooth` ops that do different things, I hope we can replace the first `smooth` with `multiply` and the second one with `quantize`.